Integration test harness: start workers too quickly

#### Is this a bug, feature request, or feedback?

Bug

#### What is the current behavior?

Intermittent bug in starting workers, e.g. https://circleci.com/gh/WallarooLabs/wallaroo/12730, where a worker is started AFIACT before the initializer worker is ready.  The output from `worker1` is:

```
--- worker1 (pid: 7478, rc: None) ->

****CLUSTERING MODE is active****
****AUTOSCALE MODE is active****
||| Resilience directory: /tmp/res-data.o13dnsep|||
Set up external channel listener on 127.0.0.1:20035
_create_control_connection: call from line 106
AUTOSCALE: Waiting for new autoscale event.
worker1 external: listening on 127.0.0.1:20035
metrics outgoing connected
worker1 control: listening on 127.0.0.1:20032
_create_control_connection: call from line 605
_create_control_connection: target_name initializer host 127.0.0.1 service 20010
Sent control message to initializer
ControlSenderConnectNotifier (to initializer): connection failed!
[... shutdown messages omitted ...]
```

... which results in this failures by the test harness:

```
cd /root/project/examples/python/word_count_with_dynamic_keys/ && \
integration_test --newline-file-sender count_this.txt \
	--validation-cmd 'python _test/validate.py --expected _expected.json --output' \
	--log-level error \
	--batch-size 1 \
	--output 'received.txt' \
	--command 'machida3 --application-module word_count_with_dynamic_keys' \
	--sink-mode newlines \
	--sink-expect 482 \
	--workers 3
2018-11-01 22:09:37,724 integration ERROR    [cluster.py:519] Encountered and error when starting up the cluster
2018-11-01 22:09:37,724 integration ERROR    [cluster.py:520] Timed out after 30 seconds while waiting for cluster to resume processing.
Traceback (most recent call last):
  File "/root/project/testing/tools/integration/cluster.py", line 510, in __init__
    self.wait_to_resume_processing(self.is_ready_timeout)
  File "/root/project/testing/tools/integration/cluster.py", line 843, in wait_to_resume_processing
    raise w.error
integration.errors.TimeoutError: Timed out after 30 seconds while waiting for cluster to resume processing.
2018-11-01 22:09:37,727 integration ERROR    [integration.py:278] Integration pipeline_test encountered an error
2018-11-01 22:09:37,727 integration ERROR    [integration.py:281] The last 10 lines of each worker were:

--- initializer (pid: 7476, rc: None) ->

ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed

--- initializer (pid: 7476, rc: None) <-
===
--- worker1 (pid: 7478, rc: None) ->

ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed
ExternalChannelConnectNotifier (worker1): server closed

--- worker1 (pid: 7478, rc: None) <-
===
--- worker2 (pid: 7484, rc: None) ->

ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed
ExternalChannelConnectNotifier (worker2): server closed

--- worker2 (pid: 7484, rc: None) <-
2018-11-01 22:09:37,727 integration ERROR    [integration_test:577] Encountered an error while running the test for 'machida3 --application-module word_count_with_dynamic_keys'
===
Traceback (most recent call last):
  File "/root/project/testing/tools/integration_test", line 559, in CLI
    persistent_data = persistent_data)
  File "/root/project/testing/tools/integration/integration.py", line 154, in pipeline_test
    persistent_data=persistent_data) as cluster:
  File "/root/project/testing/tools/integration/cluster.py", line 523, in __init__
    raise err
  File "/root/project/testing/tools/integration/cluster.py", line 510, in __init__
    self.wait_to_resume_processing(self.is_ready_timeout)
  File "/root/project/testing/tools/integration/cluster.py", line 843, in wait_to_resume_processing
    raise w.error
integration.errors.TimeoutError: Timed out after 30 seconds while waiting for cluster to resume processing.
examples/python/word_count_with_dynamic_keys/Makefile:81: recipe for target 'word_count_with_dynamic_keys_py3_test' failed
make: *** [word_count_with_dynamic_keys_py3_test] Error 1
Exited with code 2
```

#### What is the expected behavior?

Perhaps we wait until all TCP ports listened to by initializer are actually open & listening?

#### What OS and version of Wallaroo are you using?

CircleCI environment, branch for PR #2621

#### Steps to reproduce?

`make integration-tests debug=true`

--- a/lib/wallaroo/ent/network/connections.pony
+++ b/lib/wallaroo/ent/network/connections.pony
@@ -593,7 +593,7 @@ actor Connections is Cluster
         try
           _control_conns(target_name)?
         else
-          Fail(); ControlConnection(this)
+          Unreachable(); ControlConnection(this)
         end
       else
         ControlConnection(this)
@@ -601,7 +601,8 @@ actor Connections is Cluster
     _control_conns(target_name) = tcp_conn_wrapper
     _register_disposable(tcp_conn_wrapper)
     let control_notifier: TCPConnectionNotify iso =
-      ControlSenderConnectNotifier(_auth, target_name, tcp_conn_wrapper)
+      ControlSenderConnectNotifier(_auth, target_name, host, service,
+        tcp_conn_wrapper, this)
     let control_conn: TCPConnection =
       TCPConnection(_auth, consume control_notifier, host, service)
 

--- a/lib/wallaroo/ent/network/control_sender_connect_notifier.pony
+++ b/lib/wallaroo/ent/network/control_sender_connect_notifier.pony
@@ -17,26 +17,37 @@ Copyright 2018 The Wallaroo Authors.
 */
 
 use "net"
+use "time"
+use "wallaroo/core/common"
 
 class ControlSenderConnectNotifier is TCPConnectionNotify
   let _auth: AmbientAuth
-  let _worker_name: String
+  let _target_worker: WorkerName
+  let _host: String
+  let _service: String
   var _header: Bool = true
   let _tcp_conn_wrapper: ControlConnection
+  let _connections: Connections
+  let _timers: Timers = Timers
 
-  new iso create(auth: AmbientAuth, worker_name: String,
-    tcp_conn_wrapper: ControlConnection)
+  new iso create(auth: AmbientAuth, target_worker: WorkerName, host: String,
+    service: String, tcp_conn_wrapper: ControlConnection,
+    connections: Connections)
   =>
     _auth = auth
-    _worker_name = worker_name
+    _target_worker = target_worker
+    _host = host
+    _service = service
     _tcp_conn_wrapper = tcp_conn_wrapper
+    _connections = connections
 
   fun ref connected(conn: TCPConnection ref) =>
     _tcp_conn_wrapper.connected(conn)
     conn.expect(4)
     _header = true
-    @printf[I32]("ControlSenderConnectNotifier: connected to %s.\n".cstring(),
-      _worker_name.cstring())
+    @printf[I32]("ControlSenderConnectNotifier: connected to %s at %s:%s.\n"
+      .cstring(), _target_worker.cstring(), _host.cstring(),
+      _service.cstring())
 
   fun ref received(conn: TCPConnection ref, data: Array[U8] iso,
     n: USize): Bool
@@ -44,10 +55,34 @@ class ControlSenderConnectNotifier is TCPConnectionNotify
     true
 
   fun ref connect_failed(conn: TCPConnection ref) =>
-    @printf[I32]("ControlSenderConnectNotifier (to %s): connection failed!\n"
-      .cstring(), _worker_name.cstring())
+    @printf[I32](("ControlSenderConnectNotifier: connection to %s at %s:%s " +
+      " failed!\n").cstring(), _target_worker.cstring(), _host.cstring(),
+      _service.cstring())
+    let t = Timer(_Reconnect(_target_worker, _host, _service, _connections),
+      1_000_000_000)
+    _timers(consume t)
+    _tcp_conn_wrapper.closed(conn)
 
   fun ref closed(conn: TCPConnection ref) =>
-    @printf[I32]("ControlSenderConnectNotifier (to %s): server closed\n"
-      .cstring(), _worker_name.cstring())
+    @printf[I32](("ControlSenderConnectNotifier: connection to %s at %s:%s " +
+      " closed\n").cstring(), _target_worker.cstring(), _host.cstring(),
+      _service.cstring())
     _tcp_conn_wrapper.closed(conn)
+
+class _Reconnect is TimerNotify
+  let _target_worker: WorkerName
+  let _host: String
+  let _service: String
+  let _connections: Connections
+
+  new iso create(target_worker: WorkerName, host: String, service: String,
+    connections: Connections)
+  =>
+    _target_worker = target_worker
+    _host = host
+    _service = service
+    _connections = connections
+
+  fun ref apply(timer: Timer, count: U64): Bool =>
+    _connections.create_control_connection(_target_worker, _host, _service)
+    false

